### ISSUE TITLE
When Scrolled to bottom scrollOffset is not equal to scrollSize in Windows

### ISSUE THREAD

**Issue by saqlain1020** [inokawa/virtua#175]
**Describe the bug**
Cannot determine by looking at scrollOffset, scrollSize and viewportSize that the user has scrolled to bottom.
scrollOffset + viewportSize does not equal scrollSize, the sum is usually 2-15 units different.

**To Reproduce**
Scroll to bottom in a virtual list then console vlist refrence and check scrollOffset,scrollSize and viewportSize to see if by any combination you can identify if user has scrolled to bottom

**Expected behavior**
scrollOffset should be equal to scrollSize.

**Platform:**
 - OS: [Windows,]
 - Browser: [Chrome]
 - Version of this package: [0.8.2]

**Additional context**
If we can get refrence to the html element which has the scroll bar in onScroll event listener that would be make things easier.

**Comment by inokawa** [comment]
@saqlain1020 
Hi, could you tell me the value of `window.devicePixelRatio` in your chrome browser?  because it may be a problem only happens under specific dpr.

> If we can get refrence to the html element which has the scroll bar in onScroll event listener that would be make things easier.

You can access html element of scroller with passing your custom component to `components.Root` prop.

**Comment by saqlain1020** [comment]
> @saqlain1020 Hi, could you tell me the value of `window.devicePixelRatio` in your chrome browser? because it will be a problem only happens under specific dpr.
> 
> > If we can get refrence to the html element which has the scroll bar in onScroll event listener that would be make things easier.
> 
> You can access html element of scroller with passing your custom component to `components.Root` prop.

window.devicePixelRatio is 1.25

**Comment by saqlain1020** [comment]
@inokawa How can i get html element ref, basically i am thinking of checking scrollHeight for determining if scrolled to bottom.
Can you write me a snippet or something, i can use a workaround like giving class to parent and using css selector to get the element, but it would be great if we can directly get the ref.
Thanks a lot!
Your library is working pretty great, and it's straight forward.

**Comment by inokawa** [comment]
Thank you!

My recommended workaround for now is just add offset for the error like`scrollOffset > scrollSize - viewportSize - 1.5`. It's actually used in at-bottom-detection of this lib for some PCs which has dpr like 1.25 not integer.

https://github.com/inokawa/virtua/blob/146811ef8c11601441e4c0a16937efe96134af4f/src/core/store.ts#L223-L224

`VListHandle.scrollSize` returns actual scrollHeight of element currently so it may not solve the problem. However you can get the element of scroller like this:

```tsx
useEffect(() => {
  const el = document.querySelector(".foo"); 
}, [])

<VList className="foo">
  {items}
</VList>
```

I'll try to investigate why they are not equal under some dpr.

**Comment by saqlain1020** [comment]
Also there is a warning in nextjs console that useLayoutEffect should not be used in server environment, i want to use VList on client side only, so take a look at it as well, might change useLayoutEffect to useEffect

**Comment by inokawa** [comment]
> Also there is a warning in nextjs console that useLayoutEffect should not be used in server environment

I implemented not to use useLayoutEffect in server like this. Maybe it isn't working for some reason... 🤔 

https://github.com/inokawa/virtua/blob/146811ef8c11601441e4c0a16937efe96134af4f/src/react/useIsomorphicLayoutEffect.ts#L4-L5

**Comment by inokawa** [comment]
The warn will be fixed in `0.8.3`.

**Comment by inokawa** [comment]
I found that scrollSize from handle has rounded value but scrollSize internally used in this lib is not. I fixed that in `0.8.4` so you can get more precise scrollSize from handle.

I close this issue because I have no idea how to reduce the error further and I think the fix made we can detect at bottom stably now like this:
https://github.com/inokawa/virtua/blob/bd877cebd694850e13d9ac6dbe1842c376be2593/stories/basics/VList.stories.tsx#L488-L494
Of Course any improvements and information are welcome.
